oci-systemd-hook: work with OCI hooks

giuseppe opened the pull request:
make it usable from OCI spec file as:
```
"hooks": {
    "prestart": [
        {
            "path": "/usr/libexec/oci/hooks.d/oci-systemd-hook"
        }
    ],
    "poststop": [
        {
            "path": "/usr/libexec/oci/hooks.d/oci-systemd-hook"
        }
    ]
}
```
runc does not pass any argument to the process, so read all the info
from the stdin JSON file.  The mount points are read from the
config.json file in the container bundle directory.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

rhatdan commented:
@mrunalp PTAL

rhatdan commented:
Would this same path work for docker?  Can we simplify the docker patch to only pass in the hook name and rely on the oci runtime json?  Will the hooks work in a non runc oci container?

giuseppe commented:
I am not sure if it is possible to get the OCI runtime json from Docker and use only the new path.

Yes, it should work with any oci container, I have not tested it but should work even with bwrap-oci.

rhatdan commented:
Hooks should be ignored with bubblewrap. with a message stating that non priv use case does not support hooks.

mrunalp commented:
I just tested this. It works with docker 1.10.3 but doesn't work with runc latest upstream. I used the same config.json that you shared and the fedora-httpd rootfs that we have been using. 

```
Sep 16 16:01:40 localhost.localdomain.localdomain oci-systemd-hook[14027]: systemdhook <error>: Failed to move mount /testroot/fedora-httpd/rootfs//run/secrets to /tmp/runtmp.2wjgAj//secrets: Invalid argument
Sep 16 16:01:40 localhost.localdomain.localdomain oci-systemd-hook[14027]: systemdhook <error>: Failed to move /testroot/fedora-httpd/rootfs/run/secrets to /tmp/runtmp.2wjgAj: Invalid argument
Sep 16 16:01:40 localhost.localdomain.localdomain oci-systemd-hook[14027]: systemdhook <error>: Failed to set heirarchy under /run: Invalid argument
```

giuseppe commented:
I have used the version on Fedora:

$ runc --version
runc version 0.1.1
spec: 0.6.0-dev

mrunalp commented:
@giuseppe Thanks! Can we make it work with the latest runc?

giuseppe commented:
@mrunalp, fixed.  It was actually a different issue, since Docker on Fedora/RHEL always mount `/run/secrets`.  It is not true with runc.  I added a patch to check it is a mountpoint before trying to move it.
I think it worked before for me, as I have manually deleted all the files in the container `/run` directory.

Tested with Docker 1.10.3, runc 0.1.1 and git master.

mrunalp commented:
@giuseppe Nice catch. I will test this out again.

giuseppe commented:
@mrunalp have you had a chance to look at this PR?

mrunalp commented:
Will test this today. Sorry about the delay.

> On Sep 23, 2016, at 5:58 AM, Giuseppe Scrivano <notifications@github.com> wrote:
> 
> @mrunalp have you had a chance to look at this PR?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mrunalp commented:
This PR doesn't work for me with runc (I added debugs to print directory name):
```
Sep 23 14:14:26 localhost.localdomain.localdomain oci-systemd-hook[1521]: systemdhook <debug>: Mount Label parsed as: system_u:object_r:svirt_sandbox_file_t:s0:c1,c2
Sep 23 14:14:26 localhost.localdomain.localdomain oci-systemd-hook[1521]: systemdhook <error>: Failed to mkdir: /testroot/fedora-httpd/rootfs/run: No such file or directory
Sep 23 14:14:26 localhost.localdomain.localdomain oci-systemd-hook[1523]: systemdhook <debug>: Mount Label parsed as: system_u:object_r:svirt_sandbox_file_t:s0:c1,c2

```

giuseppe commented:
Thanks for the review.  I have done this test here:
```
$ cat<<EOF > Dockerfile
FROM         fedora:24
ENV container docker
RUN dnf -y install httpd; dnf clean all; systemctl enable httpd
STOPSIGNAL SIGRTMIN+3
EXPOSE 80
CMD [ "/sbin/init" ]
EOF
$ sudo docker build -t httpd .
$ sudo docker create --name httpd httpd
2571cc1f97b80d6b0e31e9d4d9eeaf39bb807d8db72fab0d4d614905b42a9a5b
$ mkdir rootfs
$ cd rootfs
$ sudo docker export httpd | tar xf -
$ cd ..
$ sudo docker rm httpd
$ wget https://gist.githubusercontent.com/giuseppe/07bcef38d1820fe50eebcfbc9cecce40/raw/809121b9c90d0f357758e8ae723735f0b8b57818/config.json
$ runc --version 
runc version 1.0.0-rc1
commit: 7ec24c513fd2e6c3d7f0c1f0a525dbb54e54ec9e
spec: 1.0.0-rc1-dev
$  sudo runc run httpd
(cut output)
```
(from another terminal to terminate it)
```
$ sudo runc kill httpd 37
```

Have you done something differently?

(I have also prepared an asciinema recording if it can be helpful: https://asciinema.org/a/0pt8uguf7419u0jnvfzkjg4kd)

mrunalp commented:
@giuseppe I have a similar setup. I'll try again by recreating my rootfs. Meanwhile, could you rebase?

giuseppe commented:
sure.  Rebased here :arrow_up:

mrunalp commented:
@giuseppe I am still seeing issues :/ Let's have a call on Monday to sync up on this.

giuseppe commented:
@mrunalp are you still seeing the same issues that you reported before?

mrunalp commented:
@giuseppe No different errors this time. Sorry, I have just been reporting errors. I'll look deeper into the issue on Monday and we can get it in.

mrunalp commented on src/systemdhook.c:
```diff
@@ -703,53 +723,103 @@ int main(int argc, char *argv[])
-	const char **config_mounts = YAJL_GET_OBJECT(v_mps)->keys;
-	unsigned config_mounts_len = YAJL_GET_OBJECT(v_mps)->len;
-	if (!strcmp("prestart", argv[1])) {
+	if ((argc > 2 && !strcmp("prestart", argv[1])) || target_pid) {
```
Can we add a comment explaining how the target_pid logic works?

mrunalp commented on src/systemdhook.c:
```diff
@@ -703,53 +723,103 @@ int main(int argc, char *argv[])
+		}
+		yajl_val v_arg0_value = YAJL_GET_ARRAY(v_args)->values[0];
+		cmd = YAJL_GET_STRING(v_arg0_value);
+	} else {
```
Can we add a comment that this is the path followed when invoked by docker?

mrunalp commented on src/systemdhook.c:
```diff
@@ -677,8 +682,23 @@ int main(int argc, char *argv[])
 	}
 	char *id = YAJL_GET_STRING(v_id);
 
+	const char *bundle_path[] = { "bundlePath", (const char *)0 };
```
I think that a comment at the beginning here explaining how this works w.r.t. arguments and configs when used with OCI and docker would help.

mrunalp commented:
@giuseppe Alright, I looked at the code closer today and reran stuff and it works fine :) I have added a few nits around more comments otherwise looks fine!

giuseppe commented:
@mrunalp thanks for the review.  I have just pushed another version that adds the missing inline comments :arrow_up:

mrunalp commented:
LGTM. Thanks!
